legate/core/communicator.py

```diff
@@ -29,24 +30,28 @@ def __init__(self, runtime: Runtime) -> None:
         self._runtime = runtime
         self._context = runtime.core_context
 
-        self._handles: dict[int, FutureMap] = {}
+        self._handles: dict[tuple[int, ProcessorRange], FutureMap] = {}
         # From launch domains to communicator future maps transformed to N-D
-        self._nd_handles: dict[Rect, FutureMap] = {}
+        self._nd_handles: dict[tuple[Rect, ProcessorRange], FutureMap] = {}
 
     def _get_1d_handle(self, volume: int) -> FutureMap:
-        if volume in self._handles:
-            return self._handles[volume]
+        proc_range = self._runtime.machine.get_processor_range()
```

Comment:
That's because `self._runtime.machine` can change; every `with some_machine:` statement modifies the internal machine state, and all machine dependent caches should use that.